### PR TITLE
Phase 1d (#235): flatten/unflatten → JAX-pytree; flat_size → vector_size

### PR DESCRIPTION
**Pull request by arob5**
## PR-1d — Consolidate 1-D (de)serialization: `flatten`/`unflatten` → JAX-pytree; `flat_size` → `vector_size`

Fourth slice of Phase 1 of #235 (follows #292, #308, #311 — all merged). **Breaking.** It closes the
duplication that #311 deliberately left behind, finishing the *numeric-vector* vs *JAX-pytree*
vocabulary split. Scope is set by **@arob5's 2026-06-22 comment on #235** (the three "PR-1d" to-dos)
plus the consumer migration they require.

> **Read `CONTRACTS.md` (on `main`) first.** This is a breaking, contract-heavy PR. Per CONTRACTS.md
> directive 1, **pin the contract decisions below *before* implementing** (and record them here);
> per directive 2, document the repurposed contracts fully in docstrings; complete the per-PR
> checklist. This is the first breaking slice since #292 — the consumer audit is the critical part.

### Scope
**(1) Repurpose `flatten`/`unflatten` to the JAX-pytree meaning.** Today the *methods*
`NumericRecord.flatten()` (`probpipe/core/_numeric_record.py:195`), `NumericRecordArray.flatten()`
(`probpipe/core/_record_array.py:440`), and their `unflatten` counterparts (`:211` / `:464`)
return/consume a **1-D vector**. That 1-D meaning now lives entirely in `to_vector`/`from_vector`
(#311). Repurpose `flatten`/`unflatten` to the **JAX-pytree** operation (`value → (leaves, aux)` /
`(leaves, aux) → value`, à la `jax.tree_util.tree_flatten`/`tree_unflatten`), defined for *any* leaf
type — not just numeric.

**(2) Collapse `to_vector`/`flatten` duplication.** Refactor `EventTemplate.to_vector` to delegate
to the repurposed `flatten` (ravel + concatenate the pytree leaves) instead of calling
`jax.tree_util.tree_leaves` inline, so leaf traversal lives in one place.

**(3) Collapse `from_vector`/`unflatten` duplication (symmetric).** Refactor
`EventTemplate.from_vector` to rebuild the numeric leaves via the repurposed `unflatten` (then weave
in `non_numeric` for the mixed case) instead of its standalone `offset`-walk, so leaf reconstruction
lives in one place.

**(4) Rename `flat_size` → `vector_size`.** **105 occurrences across 14 files** (6 source, 8 test):
3 public properties (`NumericEventTemplate.flat_size`, `NumericRecord.flat_size`,
`NumericRecordDistribution.flat_size`) + private `_flat_size` slots / `_compute_flat_size`.
`vector_size` is the length of the `to_vector`/`from_vector` 1-D representation; a *batch* is a
matrix `(*batch_shape, vector_size)`, **not** a single vector — state this in the docstring. Review
the non-mechanical `nested_flat_size` comments in `_approximate_distribution.py`.

**(5) Migrate consumers** of the old 1-D `flatten`/`unflatten` + legacy static machinery to the new
API. Audit with `git grep`; known areas: the legacy statics
`NumericRecordDistribution.flatten_value`/`unflatten_value`/`event_size`/`as_flat_distribution`
(`probpipe/core/_numeric_record_distribution.py:490–557`); the inference flatten/unflatten-sample
path (`probpipe/inference/_inference_utils.py`); `distributions/kde.py`;
`inference/_approximate_distribution.py`; `modeling/_pymc.py`. Anything calling `.flatten()` for the
1-D vector → `to_vector`; `.unflatten(vec, …)` → `from_vector`.

### Contract decisions to pin BEFORE implementing (record them in this PR)
- **`flatten()` signature:** mirror `jax.tree_util.tree_flatten` → `(leaves, aux)` with
  `unflatten(leaves, aux) → value`? (Pick the JAX-aligned form and state it precisely.)
- **Keep vs drop the methods:** the #235 comment's intent is to *keep* `flatten`/`unflatten` as the
  one place leaf-traversal lives and have `to_vector`/`from_vector` delegate to them — confirm, vs.
  callers using `jax.tree_util.*` directly.
- **`flat_size` → `vector_size`:** hard rename in-repo; add a deprecation alias only if external code
  depends on `flat_size`.
- **Legacy statics** (`flatten_value`/`unflatten_value`/`event_size`/`as_flat_distribution`): remove
  or repoint? `event_size` ≡ `vector_size`; `as_flat_distribution` / `FlatNumericRecordDistribution`
  is used by inference — likely repoint (not remove) so inference keeps working. Decide explicitly.
- **The old 1-D `flatten()` is going away** as a 1-D op — confirm every caller is migrated to
  `to_vector`/`from_vector` (no stragglers expecting a vector from `.flatten()`).

### Out of scope (do NOT do here)
- `make_batch` / `vectorize` (later, with the `Batch` ABC).
- `*Array → *Batch` rename; the `Tracked` / `Annotated` mixins (Phase 2).
- The **mixin / naming-contract discussion** in the *other* (June-18) #235 comments — that is Phase 2;
  do not start it here.

### Definition of done
- `flatten`/`unflatten` carry the JAX-pytree meaning, with docstrings that state the precise
  distinction from `to_vector`/`from_vector` (CONTRACTS.md directive 5 — keep the vocabulary split
  intact and self-contained, no PR/issue refs in docstrings).
- `to_vector`/`from_vector` delegate to `flatten`/`unflatten` (no inline `tree_leaves`, no standalone
  reconstruction); behavior unchanged — the #311 `to_vector`/`from_vector` tests pass *as-is* (the
  regression guard).
- `flat_size` → `vector_size` everywhere; `vector_size` semantics documented; `nested_flat_size`
  comments reviewed.
- All consumers migrated; full suite green; **`ruff format` and `ruff check` both clean** (blocking).
- Naming consistent with CONTRACTS.md (`vector_size`, `vec`, `value`, `non_numeric`); per-PR
  checklist completed in this description.

### Tests
- The #311 `to_vector`/`from_vector` tests pass **unchanged** (round-trip / order / errors /
  nested / multi-axis batched / `non_numeric`) — proves the delegation refactor preserved behavior.
- `flatten`/`unflatten` pytree round-trip on `Record` / `NumericRecord` / mixed templates, including
  **opaque leaves** (the pytree op is defined for any leaf type, unlike `to_vector`).
- `vector_size` returns the same values the old `flat_size` did (behavior-preserving rename),
  including the batched matrix shape `(*batch_shape, vector_size)`.
- Consumer-path regression: inference (numeric flatten/unflatten-sample), kde, approximate
  distribution still work end-to-end.

Rationale: #235 Chapter 1 + §7 and @arob5's PR-1d comment. Removes the temporary duplication #311
introduced and completes the numeric-vs-pytree vocabulary split.

---
*Draft/handoff PR — implemented by a separate session per this brief and `CONTRACTS.md`.*

---

## PINNED CONTRACT DECISIONS (resolved before coding, per CONTRACTS.md directive 1)

**1. `flatten` / `unflatten` signature — JAX-aligned, whole-tree.**
`flatten`/`unflatten` become the general JAX-pytree ops, defined on the **`Record` base**
(inherited by `NumericRecord`, `RecordArray`, `NumericRecordArray`) so they work for **any** leaf
type (numeric or opaque):
- `record.flatten() -> (leaves, aux)` ≡ `jax.tree_util.tree_flatten(record)` — the value's pytree
  leaves in canonical leaf order plus the `PyTreeDef` (`aux`). Whole-tree (recurses through nested
  records), not single-node.
- `Record.unflatten(leaves, aux) -> value` (staticmethod) ≡ `jax.tree_util.tree_unflatten(aux, leaves)`.
  Arg order is `(leaves, aux)` per the contract; `aux` is the `PyTreeDef` from `flatten`.
- Round-trip: `Record.unflatten(*record.flatten()) == record`.

**2. Keep-and-delegate; the old 1-D methods are dropped.** The old 1-D
`NumericRecord.flatten/unflatten` and `NumericRecordArray.flatten/unflatten` are **removed** (their
1-D meaning now lives solely in `to_vector`/`from_vector`). `EventTemplate.to_vector` delegates to
`value.flatten()` (ravel + concat the pytree leaves); `EventTemplate.from_vector` slices `vec` into
the ordered leaves and delegates assembly to `unflatten` (treedef derived from the template), so
leaf traversal and reconstruction each live in one place. No inline `tree_leaves`, no standalone
offset-walk reconstruction.

**3. Ergonomic `to_vector()` instance method.** `NumericRecord.to_vector()` /
`NumericRecordArray.to_vector()` are added as the instance-level convenience that replaces the old
`value.flatten()` 1-D ergonomic (they delegate to their event template's `to_vector`). The
structural op remains on `EventTemplate`; `from_vector` stays a template classmethod (no value to
bind to).

**4. `flat_size` → `vector_size` — hard rename, no alias.** All 3 public properties
(`NumericEventTemplate`, `NumericRecord`, `FlatNumericRecordDistribution`) + private
`_flat_size` slots + `_compute_flat_size` are renamed in-repo. No deprecation alias (no external
dependence known; repo-internal only). `vector_size` = length of the per-element 1-D vector; a
batch is a matrix `(*batch_shape, vector_size)` — stated in the docstrings. `nested_flat_size`
comments in `_approximate_distribution.py` reworded to `nested vector_size`.

**5. Legacy statics — repoint, do not remove.**
- `as_flat_distribution` / `FlatNumericRecordDistribution`: **kept** (the documented inference
  interop; reached via `getattr` in `_inference_utils.py`). `FlatNumericRecordDistribution.flat_size`
  → `vector_size` (the item-4 rename).
- `flatten_value` / `unflatten_value`: **kept as raw-array-aware adapters** but **repointed** onto
  `to_vector`/`from_vector` internally (they previously called the now-removed 1-D `.flatten()` /
  `.unflatten()`). They handle raw arrays and the single-field "return raw array" contract that
  `to_vector`/`from_vector` do not, so deleting them would push that handling into every view + kde.
  The 1-D logic now flows through `to_vector`/`from_vector` in one place.
- `event_size`: **kept** (not part of the item-4 `flat_size` rename; it is the base
  `NumericRecordDistribution`'s per-sample scalar count and is valid for non-flat NRDs where a
  `vector_size` property does not exist). Computed from the event template as before.

**6. All callers of the removed 1-D `.flatten()`/`.unflatten()` migrated.** Direct callers
(`NumericRecord(Array).unflatten(flat, template=...)` → `template.from_vector(flat)`;
`value.flatten()` → `value.to_vector()` / `template.to_vector(value)`) migrated across
`_record_distribution.py`, `_approximate_distribution.py`, `_bayesflow_common.py`,
`_bayesflow_likelihoods.py`, `_inference_utils.py`, `kde.py`, and the distribution-view methods.
Special care taken where `hasattr(x, "flatten")` previously discriminated 1-D — those sites now test
the Record types explicitly so the new pytree `flatten` is never mistaken for the 1-D vector.

### CONTRACTS.md per-PR checklist
- [x] Contract of every touched abstraction was clear before coding (ambiguities resolved & noted above).
- [x] New/changed contracts fully documented in docstrings (types, shapes, orderings, raises, invariants).
- [x] Code verified to obey the documented contract; tests assert it (shapes + orderings + error cases).
- [x] Docstrings describe the plan's **target** contract/terminology, not stale neighboring code; interim coexistence labeled.
- [x] Variable names consistent with the canonical table (`vector_size`, `vec`, `value`, `non_numeric`).
- [x] `CONTRACTS.md` index updated if a cross-cutting contract changed (no index change needed — locations unchanged).
- [x] `ruff format` and `ruff check` clean (both blocking in CI).

> Note on the #311 regression guard: the #311 `to_vector`/`from_vector` tests keep their behavioral
> assertions **unchanged**; only the `flat_size` → `vector_size` token is updated where they
> reference the renamed property (the rename is item 4 and unavoidably touches those references).

### CODE REVIEW

File: tests/inference/test_bayesflow_posteriors.py [hunk 1]
```diff
@@ -26,10 +26,17 @@
     condition_on,
     learn_amortized_posterior,
 )
-from probpipe.core._record_array import NumericRecordArray
 from probpipe.modeling import GenerativeLikelihood, Likelihood
 
 
+def _theta_vec(params):
```
Comment (jhuggins) [inline tests/inference/test_bayesflow_posteriors.py]: duplicated

File: tests/inference/test_bayesflow_likelihoods.py [hunk 2]
```diff
@@ -31,10 +31,18 @@
     learn_amortized_likelihood,
     learn_amortized_ratio,
 )
-from probpipe.core._record_array import NumericRecordArray
 from probpipe.inference._bayesflow_common import _adapter_field_keys
 from probpipe.modeling import GenerativeLikelihood, Likelihood
 
+
+def _theta_vec(params):
```
Comment (jhuggins) [inline tests/inference/test_bayesflow_likelihoods.py]: missing type hints

**Review by jhuggins — APPROVED** [review]
Just a couple minor items that you can fix now or in a follow-up PR

File: tests/inference/test_bayesflow_posteriors.py (hunk 1 above)
Comment (arob5) [reply tests/inference/test_bayesflow_posteriors.py]: Fixed in 0132fbb — extracted the duplicated helper into a single `tests/inference/_bayesflow_helpers.py` (`theta_vec`), now imported by both bayesflow test files.

File: tests/inference/test_bayesflow_likelihoods.py (hunk 2 above)
Comment (arob5) [reply tests/inference/test_bayesflow_likelihoods.py]: Fixed in 0132fbb — the shared `theta_vec` now has type hints (`params: Any -> Array`) and a NumPy-style docstring.

**Review by jhuggins — APPROVED** [review]
